File: lua/dnd-character.lua
local function ability(scores)
    table.sort(scores)
    return scores[2] + scores[3] + scores[4]
end

local function roll_dice()
    local t = {}
    for _ = 1, 4 do
        table.insert(t, math.random(6))
    end
    return t
end

local function modifier(input)
    return math.floor((input - 10) / 2)
end

local Character = {
    new = function(_self, name)
        local constitution = ability(roll_dice())
        return {
            name = name,
            strength = ability(roll_dice()),
            dexterity = ability(roll_dice()),
            constitution = constitution,
            intelligence = ability(roll_dice()),
            wisdom = ability(roll_dice()),
            charisma = ability(roll_dice()),
            hitpoints = 10 + modifier(constitution),
        }
    end,
}

return {
    Character = Character,
    ability = ability,
    roll_dice = roll_dice,
    modifier = modifier,
}

File: elixir/dnd_character.ex
defmodule DndCharacter do
  @type t :: %__MODULE__{
          strength: pos_integer(),
          dexterity: pos_integer(),
          constitution: pos_integer(),
          intelligence: pos_integer(),
          wisdom: pos_integer(),
          charisma: pos_integer(),
          hitpoints: pos_integer()
        }

  defstruct ~w[strength dexterity constitution intelligence wisdom charisma hitpoints]a

  @spec modifier(pos_integer()) :: integer()
  def modifier(score) do
    floor((score - 10) / 2)
  end

  @spec ability :: pos_integer()
  def ability do
    [:rand.uniform(6), :rand.uniform(6), :rand.uniform(6), :rand.uniform(6)]
    |> Enum.sort()
    |> Enum.drop(1)
    |> Enum.sum()
  end

  @spec character :: t()
  def character do
    constitution = ability()
    %DndCharacter{
      strength:  ability(),
      dexterity: ability(),
      constitution: constitution,
      intelligence: ability(),
      wisdom: ability(),
      charisma: ability(),
      hitpoints: modifier(constitution) + 10,
    }
  end
end

